x/exchange/keeper.go

```diff
@@ -123,41 +122,44 @@ func (k Keeper) fillOrderIfPossible(
 	}
 	orderBook := k.getOrderBook(ctx, matchingKind, amount.Denom, price.Denom)
 
-	// remove expired
-	orderBook.RemoveExpiredLimitOrders()
-
 	// unfilled amount
 	unfilledAmt := amount
 
 	var err sdk.Error
 
 	for i, storedOrder := range orderBook.Orders {
```

Comment:
just an idea - haven't thought about it too deeply:

being key/core logic here, we may want to have this logic done in a more functional/pure/immutable style rather than loops with breaks and calls to the bankKeeper on every loop iteration, as that is usually more conducive to safety/security analysis and ease of testing.

but probably fine for now, wouldn't worry about optimizing/changing until later on when we go into a hardcore security review phase.

what are your thoughts on those ideas?

Reply:
@musnit valid point. 

We should also take a note to not store orders within an array in an orderbook, since that produces huge orderbook values for our database and could evetually lead to performance problems. It may be much better to store orders just under their id, and reference them from orderbooks. That way, we can have them indexed by price/date within the LevelDB. 

I would leave this for now though.